packages/app-router/src/router.tsx

```diff
@@ -0,0 +1,325 @@
+import {LayoutRoot, LayoutStack, Navigation} from 'react-native-navigation';
+import {Fragment} from 'react';
+
+import {ComponentIdContext, RouteContext} from './context';
+import {
+  BottomTabRoute,
+  ComponentRoute,
+  IndexRoute,
+  RouteChildWithoutChildren,
+  Router,
+} from './types';
+
+/**
+ * Creates the initial layout object for the application.
+ *
+ * @returns {LayoutRoot} The initial layout object.
+ *
+ * @example
+ * const initialLayout = createInitialLayout();
+ * console.log(initialLayout); // { root: {} }
+ */
+function createInitialLayout(): LayoutRoot {
+  return {root: {}};
+}
+
+/**
+ * Flattens a nested array of route objects into a flat array while preserving the `stackId`
+ * of parent routes and associating it with their child routes.
+ *
+ * @param {(RootRouteObject | RouteObject)[]} routes - The array of route objects to flatten.
+ * The array can contain both `RootRouteObject` and `RouteObject`, which may have children.
+ *
+ * @param {string} [parentStackId] - An optional `stackId` passed from a parent route. If present,
+ * the `stackId` will be assigned to child routes unless the child route explicitly defines its own `stackId`.
+ *
+ * @returns {Omit<RouteObject, 'children'>[]} A flat array of route objects without the `children` property,
+ * each with the appropriate `stackId` inherited from parent routes when necessary.
+ *
+ * @remarks
+ * This function works recursively, ensuring that any nested routes are flattened into a single-level array.
+ * The `stackId` is passed down the route hierarchy, so that child routes inherit the `stackId` from their parent
+ * if they don't define their own. Routes are represented without their `children` property after flattening.
+ *
+ * @example
+ * const routes: RouteObject[] = [
+ *   {
+ *     name: 'Home',
+ *     path: '/',
+ *     Component: HomeComponent,
+ *     children: [
+ *       {
+ *         name: 'Profile',
+ *         path: '/profile',
+ *         Component: ProfileComponent,
+ *         stackId: 'mainStack',
+ *       },
+ *     ],
+ *   },
+ * ];
+ *
+ * const flattened = flatten(routes);
+ *
+ * // Output:
+ * // [
+ * //   { name: 'Home', path: '/', Component: HomeComponent },
+ * //   { name: 'Profile', path: '/profile', Component: ProfileComponent, stackId: 'mainStack' }
+ * // ]
+ */
+function flatten(
+  routes: IndexRoute[],
+  parentStackId?: string,
+): RouteChildWithoutChildren[] {
+  return routes.flatMap(route => {
+    // Extract children and the rest of the route properties
+    const {children, ...routeProps} = route as any;
```

Comment:
I think it's because `RouteChildWithoutChildren` type also includes the `ActionRoute`, we can handle it like this

```ts
if (route.type === 'action') return;
```

Reply:
updated.